test/Tester/StorageFacet/Feature.Abstractions/IStorageFeature.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System.Threading.Tasks;
+
+namespace Tester.StorageFacet.Abstractions
+{
+    /// <summary>
+    /// Primary storage feature interface.
+    ///  This is the actual functionality the users need.
+    /// </summary>
+    /// <typeparam name="TState"></typeparam>
+    public interface IStorageFeature<TState>
+    {
+        TState State { get; set; }
+
+        Task Save();
+
+        // Test calls - used to verify facet wiring works
+        string Name { get; }
+        string GetExtendedInfo();
+    }
+
+    /// <summary>
+    /// Feature configuration information which application layer can provide to the
+    ///  feature per instance (by grain type if using attributes).
+    /// </summary>
+    public interface IStorageFeatureConfig
+    {
+        string StateName { get; }
+    }
+
+    /// <summary>
+    /// Feature configuration utility class
+    /// </summary>
+    public class StorageFeatureConfig : IStorageFeatureConfig
+    {
+        public StorageFeatureConfig(string stateName)
+        {
+            this.StateName = stateName;
+        }
+
+        public string StateName { get; }
+    }
+
+    /// <summary>
+    /// Creates a storage feature from a configuration
+    /// </summary>
+    public interface IStorageFeatureFactory
+    {
+        IStorageFeature<TState> Create<TState>(IStorageFeatureConfig config);
```

Comment:
You can register IStorageFeatureFactory<> in the container, correct, but you can't resolve it as a named service as that takes resolving a generic with an imbedded generic.
This is the technical difficulty with what generic factories, the rest of the issues were more subjective.  I don't think the tradeoffs favor generic factories, but unless we deal with the technical limitations of containers to resolve imbedded generics, other debatable points are moot.

Regarding the MakeGenericMethod in StorageFeatureFactory, that one is an implementation detail of the factory, so I wasn't considering that one, since it's trivial to work around without moving to a generic factory.  I was referring to the one used to call the factory, in StorageFeatureAttributeMapper.  The general pattern builds no generic types or methods per grain activation, only the implementation's factories do that, and I'm not terribly concerned about the implementation of test features.